sklearn/utils/__init__.py

```diff
@@ -319,60 +271,76 @@ def _check_key_type(key, superclass):
     return False
 
 
-def _safe_indexing_column(X, key):
-    """Get feature column(s) from input data X.
+def safe_indexing(X, indices, axis=0):
+    """Return rows, items or columns of X using indices.
 
-    Supported input types (X): numpy arrays, sparse arrays and DataFrames.
+    Parameters
+    ----------
+    X : array-like, sparse-matrix, list, pandas.DataFrame, pandas.Series
+        Data from which to sample rows, items or columns. `list` are only
+        supported when `axis=0`.
+    indices : bool, int, str, slice, array-like
+        - If `axis=0`, boolean and integer array-like, integer slice,
+          and scalar integer are supported.
+        - If `axis=1`:
+            - to select a single column, `indices` can be of `int` type for
+              all `X` types and `str` only for dataframe. The selected subset
+              will be 1D, unless `X` is a sparse matrix in which case it will
+              be 2D.
+            - to select multiples columns, `indices` can be one of the
+              following: `list`, `array`, `slice`. The type used in
+              these containers can be one of the following: `int`, 'bool' and
+              `str`. However, `str` is only supported when `X` is a dataframe.
+              The selected subset will be 2D.
+    axis : int, default=0
+        The axis along which `X` will be subsampled. `axis=0` will select
+        rows while `axis=1` will select columns.
 
-    Supported key types (key):
-    - scalar: output is 1D;
-    - lists, slices, boolean masks: output is 2D.
+    Returns
+    -------
+    subset
+        Subset of X on axis 0 or 1.
 
-    Supported key data types:
-    - integer or boolean mask (positional):
-        - supported for arrays, sparse matrices and dataframes.
-    - string (key-based):
-        - only supported for dataframes;
-        - So no keys other than strings are allowed (while in principle you
-          can use any hashable object as key).
+    Notes
+    -----
+    CSR, CSC, and LIL sparse matrices are supported. COO sparse matrices are
+    not supported.
     """
-    # check that X is a 2D structure
-    if X.ndim != 2:
+    if indices is None:
+        return X
+
+    if axis not in (0, 1):
+        raise ValueError(
+            "'axis' should be either 0 (to index rows) or 1 (to index "
+            " column). Got {} instead.".format(axis)
+        )
+
+    if axis == 0 and _check_key_type(indices, str):
+        raise ValueError(
+            "'axis=0' only support integer or boolean array-like, slice with "
```

Comment:
can't we just say "String indexing is not supported for axis=0"?

Reply:
yes, it will much simpler